run-jsc-stress-tests --treat-failing-as-flaky should also limit total runtime

aoikonomopoulos opened the pull request:
#### 4777f045be98e46519905adaffccaf1e081d0856
<pre>
run-jsc-stress-tests --treat-failing-as-flaky should also limit total runtime
<a href="https://bugs.webkit.org/show_bug.cgi?id=290906">https://bugs.webkit.org/show_bug.cgi?id=290906</a>

Reviewed by Justin Michaud.

When we treat failing tests as flaky we retry them up to a
user-specified number of iterations, until it&apos;s clear whether they can
make the pass percentage. This works well for tests that finish quickly
but gets us in pathological behavior when a test hangs. We then run it
until it hits the timeout multiple times, needlessly prolonguing the
time it tests to run tests in the buildbot.

Add a cap on the number of seconds we can spend retrying and refuse to
start a new iteration if it has been exceeded.

* Tools/Scripts/run-javascriptcore-tests:
* Tools/Scripts/run-jsc-stress-tests:
* Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:

Canonical link: <a href="https://commits.webkit.org/294423@main">https://commits.webkit.org/294423@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f39120f90236454464abb0eb13baca6a7f314db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9767 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104965 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50417 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101880 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27920 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75996 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33085 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15084 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90156 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56361 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49789 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92494 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8228 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107325 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98445 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26950 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19680 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84950 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27312 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86361 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84476 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21895 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29152 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6872 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20762 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26886 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122071 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26697 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34081 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28258 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
LGTM, but I would like for someone more familiar with automation to take a look just in case this might break remote testing (cc @dewei-zhu @xw)

gsnedders commented on Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:
```diff
@@ -50,6 +51,9 @@ def setMaxIterations(value)
         @maxIterations = value
     end
+    def updateRetryFuel
+        raise "This method needs to be overriden"
```
Nit:

```suggestion
        raise "This method needs to be overridden"
```

gsnedders commented on Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:
```diff
@@ -127,6 +131,10 @@ def loop
                 break
             end
+            if not @retryFuel.nil? and @retryFuel < 0
+                log("Ran out of retry fuel")
```
This feels very opaque as a message; I wouldn't expect most people to know what that means. I suspect something like "exceeded maximum retry time" is a better message here?

aoikonomopoulos replied:
Fixed, thanks!

gsnedders commented on Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:
```diff
@@ -40,6 +40,7 @@ def initialize(runlist, serialPlans, remoteHosts, iterationLimits,
         end
         @maxIterations = setMaxIterations(initialMaxAllowedIterations(runlist))
         @logStream = logStream
+        @retryFuel = nil
```
I do wonder if we should just choose a clearer name here, something closer to what's in `--help`.

Maybe just `remainingRetrySeconds`?

aoikonomopoulos replied:
"Retry fuel" was meant to be internal terminology, but there's no point in having two names for the same concept. Going with `remainingRetrySeconds` as you suggest.

gsnedders commented on Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:
```diff
@@ -163,11 +171,25 @@ def loop
+        return if @treatFailingAsFlaky.nil? or @treatFailingAsFlaky.maxRetrySeconds.nil?
+        if @retryFuel.nil?
+            @retryFuel = @treatFailingAsFlaky.maxRetrySeconds
+            @prevRetry = Time.now
```
Pedantically, we should be using `Process.clock_gettime(Process::CLOCK_MONOTONIC)` when we're measuring the duration (because the system clock doesn't always move forwards).

justinmichaud replied:
Nice catch!

aoikonomopoulos replied:
That's not pedantic at all! `Process.clock_gettime` seems to be available at least as far back as ruby 2.2 so should hopefully be safe to use.

gsnedders reviewed (commented):
The other two would be better than me to comment on this, but from what I can see, our remote device testing is all using `--no-retry` today, which means that this wouldn't change anything.

aoikonomopoulos committed:
run-jsc-stress-tests --treat-failing-as-flaky should also limit total runtime
https://bugs.webkit.org/show_bug.cgi?id=290906

Reviewed by Justin Michaud.

When we treat failing tests as flaky we retry them up to a
user-specified number of iterations, until it's clear whether they can
make the pass percentage. This works well for tests that finish quickly
but gets us in pathological behavior when a test hangs. We then run it
until it hits the timeout multiple times, needlessly prolonguing the
time it tests to run tests in the buildbot.

Add a cap on the number of seconds we can spend retrying and refuse to
start a new iteration if it has been exceeded.

* Tools/Scripts/run-javascriptcore-tests:
* Tools/Scripts/run-jsc-stress-tests:
* Tools/Scripts/webkitruby/jsc-stress-test/executor.rb:

Canonical link: https://commits.webkit.org/294423@main
